<a href="https://colab.research.google.com/github/SAULIZA/prediccion_de_ventas/blob/main/Prediccion_de_ventas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PREDICCIONES DE VENTAS Proyecto1- Parte2(base)**

*   Elemento de lista
*   Elemento de lista


## Creado por: Saúl Iza

* Fecha creado: 24/02/2023
* Fecha modificada: 24/02/2023
* Descripción: proyecto


In [ ]:
import pandas as pd
filename ="/content/drive/MyDrive/sales_predictions.csv"
df = pd.read_csv(filename)
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

1) ¿Cuántas filas y columnas?

In [ ]:
df.shape

(8523, 12)

El data frame tiene 8523 filas y 12 columnas

2) ¿Cuáles son los tipos de datos en cada variable?

In [ ]:
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

Se pueden observar que hay variables de tipo float y de tipo object

3) ¿Hay duplicados? Si es el caso, eliminen algunos duplicados.

In [ ]:
df.duplicated().sum()

0

No existen duplicados

4) Identifiquen los valores faltantes.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Existen valores faltantes en Item_Weight y Outlet_Size

5) Decidan cómo abordar los valores faltantes y cómo hacerlo. (Esto requiere sus criterios, así que expliquen su elección).

5.1) Columna Item_Weight

In [ ]:
#Primero trabajamnos en la columna Item weight (peso del artículo)
df["Item_Weight"].isna().sum()

1463

Existen 1463 datos faltantes, esta variable no se puede eliminar ya que es un porcentaje de datos pequeño que falta. Por otro lado no eliminaría esta columna ya que el peso puede ser una característica importante para analizar la logísitca.

In [ ]:
# probablemente haya una correlación entre el peso y el tipo de artículo
df.groupby("Item_Type")["Item_Weight"].mean()

Item_Type
Baking Goods             12.277108
Breads                   11.346936
Breakfast                12.768202
Canned                   12.305705
Dairy                    13.426069
Frozen Foods             12.867061
Fruits and Vegetables    13.224769
Hard Drinks              11.400328
Health and Hygiene       13.142314
Household                13.384736
Meat                     12.817344
Others                   13.853285
Seafood                  12.552843
Snack Foods              12.987880
Soft Drinks              11.847460
Starchy Foods            13.690731
Name: Item_Weight, dtype: float64

In [ ]:
df.groupby("Item_Type")["Item_Weight"].min()

Item_Type
Baking Goods             4.880
Breads                   4.635
Breakfast                6.425
Canned                   4.615
Dairy                    4.805
Frozen Foods             4.555
Fruits and Vegetables    5.460
Hard Drinks              4.610
Health and Hygiene       5.175
Household                5.030
Meat                     5.150
Others                   5.500
Seafood                  5.365
Snack Foods              5.095
Soft Drinks              4.590
Starchy Foods            6.695
Name: Item_Weight, dtype: float64

In [ ]:
df.groupby("Item_Type")["Item_Weight"].max()

Item_Type
Baking Goods             20.85
Breads                   20.85
Breakfast                21.10
Canned                   21.35
Dairy                    20.70
Frozen Foods             20.85
Fruits and Vegetables    21.35
Hard Drinks              19.70
Health and Hygiene       21.25
Household                21.25
Meat                     21.25
Others                   20.50
Seafood                  20.75
Snack Foods              21.25
Soft Drinks              20.75
Starchy Foods            21.20
Name: Item_Weight, dtype: float64

El peso del artículo no depende del tipo de artículo


In [ ]:
df.groupby("Outlet_Type")["Item_Weight"].mean()

Outlet_Type
Grocery Store        12.913153
Supermarket Type1    12.849509
Supermarket Type2    12.873346
Supermarket Type3          NaN
Name: Item_Weight, dtype: float64

El peso del artículo tampoco depende del tipo de outlet. Además el supermercado 3 no ha registrado el peso de sus artículos

In [ ]:
df.groupby("Outlet_Size")["Item_Weight"].mean()

Outlet_Size
High      13.006148
Medium    12.895420
Small     12.848075
Name: Item_Weight, dtype: float64

Debido a que no se encuentra una coorelación de la columna de peso de artículo con el resto de columnas, podríamos optar por rellenar los vacios con un promedio de peso.

In [ ]:
mean_weight = df["Item_Weight"].mean()#calculo del promedio
df["Item_Weight"].fillna(mean_weight, inplace =True)# rellanamos los valores Na con el promedio
df.isna().sum() #comprobamos

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

5.2) Columna Outlet_Size

In [ ]:
df["Outlet_Size"].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

No se puede ver una variable mayoritaria con respecto al tamaño del outlet

In [ ]:
df["Outlet_Location_Type"].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

No hay correlación entre el tipo de ubicación y el tamaño del outlet

In [ ]:
df["Outlet_Identifier"].value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

In [ ]:
outl_size = ["Medium", "Small", "High"]
out = ["OUT027", "OUT013", "OUT049", "OUT046", "OUT035", "OUT045", "OUT018", "OUT017", "OUT010", "OUT019"]
for x in range(len(outl_size)):
  for y in range(len(out)):
    comun = (df["Outlet_Size"] == outl_size[x]) & (df["Outlet_Identifier"] == out[y])
    print("el tamaño de la tienda",outl_size[x], "con identificador",out[y], "tiene en comun", comun.sum())

el tamaño de la tienda Medium con identificador OUT027 tiene en comun 935
el tamaño de la tienda Medium con identificador OUT013 tiene en comun 0
el tamaño de la tienda Medium con identificador OUT049 tiene en comun 930
el tamaño de la tienda Medium con identificador OUT046 tiene en comun 0
el tamaño de la tienda Medium con identificador OUT035 tiene en comun 0
el tamaño de la tienda Medium con identificador OUT045 tiene en comun 0
el tamaño de la tienda Medium con identificador OUT018 tiene en comun 928
el tamaño de la tienda Medium con identificador OUT017 tiene en comun 0
el tamaño de la tienda Medium con identificador OUT010 tiene en comun 0
el tamaño de la tienda Medium con identificador OUT019 tiene en comun 0
el tamaño de la tienda Small con identificador OUT027 tiene en comun 0
el tamaño de la tienda Small con identificador OUT013 tiene en comun 0
el tamaño de la tienda Small con identificador OUT049 tiene en comun 0
el tamaño de la tienda Small con identificador OUT046 tiene e

No existe una correlación entre el tamaño del outlet y el identificador del outlet

In [ ]:
outl_size = ["Medium", "Small", "High"]
out_loc_t = ["Tier 1", "Tier 2", "Tier 3"]
for x in range(len(outl_size)):
  for y in range(len(out_loc_t)):
    comun1 = (df["Outlet_Size"] == outl_size[x]) & (df["Outlet_Location_Type"] == out_loc_t[y])
    print("el tamaño de la tienda",outl_size[x], "con tipo de locacion",out_loc_t[y], "tiene en comun", comun1.sum())

el tamaño de la tienda Medium con tipo de locacion Tier 1 tiene en comun 930
el tamaño de la tienda Medium con tipo de locacion Tier 2 tiene en comun 0
el tamaño de la tienda Medium con tipo de locacion Tier 3 tiene en comun 1863
el tamaño de la tienda Small con tipo de locacion Tier 1 tiene en comun 1458
el tamaño de la tienda Small con tipo de locacion Tier 2 tiene en comun 930
el tamaño de la tienda Small con tipo de locacion Tier 3 tiene en comun 0
el tamaño de la tienda High con tipo de locacion Tier 1 tiene en comun 0
el tamaño de la tienda High con tipo de locacion Tier 2 tiene en comun 0
el tamaño de la tienda High con tipo de locacion Tier 3 tiene en comun 932


In [ ]:
df.groupby("Outlet_Size")["Item_Outlet_Sales"].mean() # veamos si hay una relación entre el tamaño de la tienda y las ganancias de ventas


Outlet_Size
High      2298.995256
Medium    2681.603542
Small     1912.149161
Name: Item_Outlet_Sales, dtype: float64

tampoco hay una correlación entre el tamaño del outlet y las ventas

Debido a que no se encuentran correlación alguna, y no hay una tendencia en la variable tamaño del outlet, podríamos eliminarla, además esta característica pienso que no es importante a la hora de predecir las ventas.

In [ ]:
df.drop(columns="Outlet_Size", inplace = True)
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Location_Type        Outlet_Type  \
0                       1999               Tier 1  Supermarket Type1   
1                       2009               Tier 3  Supermarket Type2   
2                       1999               Tier 1  Supermarket Type1   
3                       1998               Tier 3      Grocery Store   
4                       1987               Tier 3  Supermarket Type1   

   Item_Outlet_Sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           732.3800  
4           994.7052

In [ ]:
#confirmamos que no hay valores faltantes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8523 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Location_Type       8523 non-null   object 
 9   Outlet_Type                8523 non-null   object 
 10  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(6)
memory usage: 732.6+ KB


7) Encuentren y arreglen alguna categoría inconsistente de datos (example: fix cat, Cat, and cats so that they are consistent) 

In [ ]:
df["Item_Fat_Content"].value_counts() #cantidad de valores en contenido de grasa

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

Se observan categorías inconsistentes, en la variable de contenido de grasa. Supondremos que Low Fat es igual que LF y también a low fat, además que Regular es igual que reg.

In [ ]:
df["Item_Fat_Content"] = df["Item_Fat_Content"].replace("LF", "Low Fat")
df["Item_Fat_Content"] = df["Item_Fat_Content"].replace("low fat", "Low Fat")
df["Item_Fat_Content"] = df["Item_Fat_Content"].replace("reg", "Regular")

In [ ]:
df["Item_Fat_Content"].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

8) Para cualquier columna numérica, obtengan las estadísticas resumidas para cada uno (mínimo, máximo y media)

In [ ]:
#Una columna que me parece muy interesante de analizar son las ventas
df["Item_Outlet_Sales"].describe().round(2)

count     8523.00
mean      2181.29
std       1706.50
min         33.29
25%        834.25
50%       1794.33
75%       3101.30
max      13086.96
Name: Item_Outlet_Sales, dtype: float64

El promedio de las ventas es de 2181.29 dólares, pero los datso de ventas son bastante dispersos, ya que se puede ver una desviación estandar de 1706,5